common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsOAuth2Strategy.java

```diff
@@ -352,6 +362,42 @@ protected void validateTokenRequest(MicrosoftStsTokenRequest request) {
         //TODO implement
     }
 
+    @Override
+    protected HttpResponse performTokenRequest(final MicrosoftStsTokenRequest request) throws IOException, ClientException {
+        final String methodName = ":performTokenRequest";
+        HttpResponse response =  super.performTokenRequest(request);
+
+        if (response.getStatusCode() == HttpURLConnection.HTTP_UNAUTHORIZED
+                && response.getHeaders() != null
+                && response.getHeaders().containsKey(CHALLENGE_REQUEST_HEADER)) {
+            // Received the device certificate challenge request. It is sent in 401 header.
+            final String requestBody = ObjectMapper.serializeObjectToFormUrlEncoded(request);
```

Comment:
This logic from line 374 to 396 can be moved to a helper method `performPkeyAuthRequest` for readability.